File: scripts/generate_workflow_pdfs.py
from pathlib import Path

from reportlab.lib import colors
from reportlab.lib.enums import TA_LEFT
from reportlab.lib.pagesizes import A4
from reportlab.lib.styles import ParagraphStyle, getSampleStyleSheet
from reportlab.lib.units import inch
from reportlab.platypus import (
    PageBreak,
    Paragraph,
    SimpleDocTemplate,
    Spacer,
    Table,
    TableStyle,
)


ROOT = Path(__file__).resolve().parents[1]
OUT = ROOT / "output" / "pdf"
OUT.mkdir(parents=True, exist_ok=True)

styles = getSampleStyleSheet()
styles.add(
    ParagraphStyle(
        name="WorkflowTitle",
        parent=styles["Title"],
        fontName="Helvetica-Bold",
        fontSize=24,
        leading=28,
        textColor=colors.HexColor("#0f172a"),
        spaceAfter=12,
    )
)
styles.add(
    ParagraphStyle(
        name="WorkflowBody",
        parent=styles["BodyText"],
        fontName="Helvetica",
        fontSize=10.5,
        leading=15,
        alignment=TA_LEFT,
        textColor=colors.HexColor("#334155"),
        spaceAfter=8,
    )
)
styles.add(
    ParagraphStyle(
        name="WorkflowSection",
        parent=styles["Heading2"],
        fontName="Helvetica-Bold",
        fontSize=15,
        leading=18,
        textColor=colors.HexColor("#0f172a"),
        spaceBefore=10,
        spaceAfter=6,
    )
)


def header_footer(canvas, doc):
    canvas.saveState()
    canvas.setStrokeColor(colors.HexColor("#cbd5e1"))
    canvas.setLineWidth(0.6)
    canvas.line(doc.leftMargin, A4[1] - 38, A4[0] - doc.rightMargin, A4[1] - 38)
    canvas.setFont("Helvetica-Bold", 9)
    canvas.setFillColor(colors.HexColor("#0f172a"))
    canvas.drawString(doc.leftMargin, A4[1] - 28, "ProjectMind Workflow Pack")
    canvas.setFont("Helvetica", 8)
    canvas.setFillColor(colors.HexColor("#64748b"))
    canvas.drawRightString(A4[0] - doc.rightMargin, 18, f"Page {doc.page}")
    canvas.restoreState()


def table(data, widths):
    tbl = Table(data, colWidths=widths, hAlign="LEFT")
    tbl.setStyle(
        TableStyle(
            [
                ("BACKGROUND", (0, 0), (-1, 0), colors.HexColor("#0ea5e9")),
                ("TEXTCOLOR", (0, 0), (-1, 0), colors.white),
                ("FONTNAME", (0, 0), (-1, 0), "Helvetica-Bold"),
                ("FONTSIZE", (0, 0), (-1, -1), 9),
                ("GRID", (0, 0), (-1, -1), 0.4, colors.HexColor("#cbd5e1")),
                ("ROWBACKGROUNDS", (0, 1), (-1, -1), [colors.white, colors.HexColor("#f8fafc")]),
                ("VALIGN", (0, 0), (-1, -1), "TOP"),
                ("LEFTPADDING", (0, 0), (-1, -1), 6),
                ("RIGHTPADDING", (0, 0), (-1, -1), 6),
                ("TOPPADDING", (0, 0), (-1, -1), 6),
                ("BOTTOMPADDING", (0, 0), (-1, -1), 6),
            ]
        )
    )
    return tbl


def story_intro():
    story = []
    story.append(Paragraph("ProjectMind AI Analysis Workflow", styles["WorkflowTitle"]))
    story.append(
        Paragraph(
            "This document explains how the app works end to end: upload a project PDF, extract text, build the project model, refresh every page, and let the AI Copilot answer from the same live state.",
            styles["WorkflowBody"],
        )
    )
    story.append(Paragraph("What the judges should understand", styles["WorkflowSection"]))
    story.append(
        Paragraph(
            "The app is not a static mockup. Uploaded PDFs are parsed by the backend, the document content is converted into structured project insights, and the dashboard, risks, commissioning, tracking, analytics, and chat all re-render from that same model.",
            styles["WorkflowBody"],
        )
    )
    return story


def build_overview_pdf(path):
    doc = SimpleDocTemplate(str(path), pagesize=A4, rightMargin=36, leftMargin=36, topMargin=54, bottomMargin=36)
    story = story_intro()
    story.append(Spacer(1, 8))
    story.append(Paragraph("System architecture", styles["WorkflowSection"]))
    story.append(
        table(
            [
                ["Layer", "Role"],
                ["Frontend", "Handles upload, loading states, live dashboards, chat UI, and PDF export actions."],
                ["Backend", "Accepts PDFs, extracts text, builds derived project data, stores state, and serves API responses."],
                ["AI layer", "Uses Gemini when configured, otherwise falls back to a project-aware local responder."],
                ["State store", "Persists uploaded documents, messages, and the derived project model for the whole app."],
            ],
            [100, 360],
        )
    )
    story.append(Spacer(1, 8))
    story.append(Paragraph("Why this matters", styles["WorkflowSection"]))
    story.append(
        Paragraph(
            "Every major screen reads from the same backend state. That means one upload can change the executive dashboard, the schedule slip story, the commissioning failure forecast, and the shipment tracking board all at once.",
            styles["WorkflowBody"],
        )
    )
    story.append(PageBreak())
    story.append(Paragraph("Upload-to-insight flow", styles["WorkflowTitle"]))
    story.append(
        table(
            [
                ["Step", "What happens"],
                ["1. Upload", "User drops one or more PDFs into the Documents page."],
                ["2. Parse", "Backend extracts text and stores a preview plus full text snippet."],
                ["3. Analyze", "Rules infer schedule risk, commissioning risk, tracking signals, and analytics."],
                ["4. Persist", "Updated documents and project model are written to local backend state."],
                ["5. Refresh UI", "Dashboard, chat, and all other pages re-read the updated live model."],
                ["6. Export", "The current project state can be exported as a real PDF summary."],
            ],
            [90, 370],
        )
    )
    story.append(Spacer(1, 8))
    story.append(Paragraph("Judge demo script", styles["WorkflowSection"]))
    story.append(
        Paragraph(
            "Show the upload, wait for the visible analyzing state, then point out the 'Insights ready' message and walk through the pages to prove the document changed the app. Finish by opening Copilot and asking a question about the uploaded content.",
            styles["WorkflowBody"],
        )
    )
    doc.build(story, onFirstPage=header_footer, onLaterPages=header_footer)


def build_commissioning_pdf(path):
    doc = SimpleDocTemplate(str(path), pagesize=A4, rightMargin=36, leftMargin=36, topMargin=54, bottomMargin=36)
    story = story_intro()
    story.append(Paragraph("Commissioning Assistant logic", styles["WorkflowSection"]))
    story.append(
        Paragraph(
            "This feature reads a commissioning test script or IST checklist and compares it with extracted document content, especially punch-list items and incomplete installs.",
            styles["WorkflowBody"],
        )
    )
    story.append(
        table(
            [
                ["Checklist item", "Likely result", "Reason"],
                ["Temporary power energization", "Likely fail", "UPS delivery delay blocks the test."],
                ["Switchgear functional test", "Likely fail", "Panel room install is incomplete."],
                ["BMS integration test", "At risk", "I/O verification is still pending."],
                ["Fire alarm interface", "Pass", "No blocking issue found."],
            ],
            [150, 90, 180],
        )
    )
    story.append(Spacer(1, 8))
    story.append(Paragraph("Main use", styles["WorkflowSection"]))
    story.append(
        Paragraph(
            "It helps the team avoid wasted test windows by warning that a test is likely to fail before anyone travels to site or starts the IST sequence.",
            styles["WorkflowBody"],
        )
    )
    doc.build(story, onFirstPage=header_footer, onLaterPages=header_footer)


def build_tracking_pdf(path):
    doc = SimpleDocTemplate(str(path), pagesize=A4, rightMargin=36, leftMargin=36, topMargin=54, bottomMargin=36)
    story = story_intro()
    story.append(Paragraph("Geospatial Tracking logic", styles["WorkflowSection"]))
    story.append(
        Paragraph(
            "This feature visualizes equipment movement and shipment status. It is a polished layer for showing where long-lead items are and which downstream task they affect.",
            styles["WorkflowBody"],
        )
    )
    story.append(
        table(
            [
                ["Equipment", "Current state", "Impact"],
                ["UPS", "In transit", "Delays installation and energization."],
                ["Transformer", "At dispatch hub", "Can push procurement-critical tasks."],
                ["Chiller", "On site", "No major downstream exposure."],
                ["Switchgear", "Customs clearance", "May affect commissioning readiness."],
            ],
            [120, 120, 190],
        )
    )
    story.append(Spacer(1, 8))
    story.append(Paragraph("Main use", styles["WorkflowSection"]))
    story.append(
        Paragraph(
            "The map is there to make the supply chain easy to understand at a glance, but the real power is that each shipment feeds back into the same project risk engine.",
            styles["WorkflowBody"],
        )
    )
    doc.build(story, onFirstPage=header_footer, onLaterPages=header_footer)


def build_judge_guide(path):
    doc = SimpleDocTemplate(str(path), pagesize=A4, rightMargin=36, leftMargin=36, topMargin=54, bottomMargin=36)
    story = story_intro()
    story.append(Paragraph("Judge demo checklist", styles["WorkflowSection"]))
    story.append(
        table(
            [
                ["Action", "What the judge sees"],
                ["Upload a PDF", "Spinner / analyzing state appears on screen."],
                ["Wait for completion", "Insights ready message confirms the AI has processed the file."],
                ["Open Dashboard", "Metrics and risk cards update from extracted content."],
                ["Open Commissioning", "Likely failures are flagged before testing."],
                ["Open Copilot", "Ask a question and get a project-aware response."],
                ["Export PDF", "A real PDF summary downloads from the backend."],
            ],
            [120, 320],
        )
    )
    story.append(Spacer(1, 8))
    story.append(Paragraph("Best presentation flow", styles["WorkflowSection"]))
    story.append(
        Paragraph(
            "Start with a clean upload, pause on the analyzing state, then show the data-driven pages changing. End with chat and export so it is obvious the system is connected end to end.",
            styles["WorkflowBody"],
        )
    )
    doc.build(story, onFirstPage=header_footer, onLaterPages=header_footer)


def main():
    build_overview_pdf(OUT / "app_workflow_overview.pdf")
    build_commissioning_pdf(OUT / "commissioning_assistant_workflow.pdf")
    build_tracking_pdf(OUT / "geospatial_tracking_workflow.pdf")
    build_judge_guide(OUT / "judge_demo_guide.pdf")


if __name__ == "__main__":
    main()

File: scripts/generate_workflow_pdfs.mjs
import fs from "node:fs"
import path from "node:path"
import { fileURLToPath } from "node:url"
import PDFDocument from "pdfkit"

const __dirname = path.dirname(fileURLToPath(import.meta.url))
const root = path.resolve(__dirname, "..")
const outDir = path.join(root, "output", "pdf")
fs.mkdirSync(outDir, { recursive: true })

function createPdf(filename, title, sections) {
  const doc = new PDFDocument({ size: "A4", margin: 44 })
  const stream = fs.createWriteStream(path.join(outDir, filename))
  doc.pipe(stream)

  doc.font("Helvetica-Bold").fontSize(22).fillColor("#0f172a").text(title)
  doc.moveDown(0.4)
  doc.font("Helvetica").fontSize(11).fillColor("#334155").text(
    "ProjectMind Workflow Pack",
    { underline: false },
  )
  doc.moveDown(0.8)

  for (const section of sections) {
    doc.font("Helvetica-Bold").fontSize(14).fillColor("#0f172a").text(section.heading)
    doc.moveDown(0.35)
    doc.font("Helvetica").fontSize(10.5).fillColor("#334155")
    for (const paragraph of section.paragraphs || []) {
      doc.text(paragraph, { width: 503 })
      doc.moveDown(0.35)
    }
    if (section.bullets?.length) {
      doc.moveDown(0.15)
      for (const bullet of section.bullets) {
        doc.text(`• ${bullet}`, { indent: 10, width: 493 })
      }
    }
    if (section.table?.length) {
      doc.moveDown(0.35)
      const widths = section.tableWidths || [180, 320]
      const startX = doc.x
      let y = doc.y
      section.table.forEach((row, rowIndex) => {
        let x = startX
        const rowHeight = rowIndex === 0 ? 22 : 28
        row.forEach((cell, index) => {
          doc.rect(x, y, widths[index], rowHeight).stroke("#cbd5e1")
          doc.fillColor(rowIndex === 0 ? "#ffffff" : "#334155")
            .font(rowIndex === 0 ? "Helvetica-Bold" : "Helvetica")
            .fontSize(9)
            .text(String(cell), x + 6, y + 6, { width: widths[index] - 12, height: rowHeight - 8 })
          x += widths[index]
        })
        y += rowHeight
      })
      doc.y = y + 10
    }
    doc.moveDown(0.6)
  }

  doc.end()
  return new Promise((resolve) => stream.on("finish", resolve))
}

await createPdf("app_workflow_overview.pdf", "ProjectMind AI Analysis Workflow", [
  {
    heading: "What this app does",
    paragraphs: [
      "ProjectMind is a live prototype that reads uploaded project PDFs, extracts text, derives a structured project model, and refreshes the dashboard, risk board, commissioning assistant, tracking map, analytics, and AI chat from the same state.",
      "It is intentionally built so the upload is not cosmetic. The backend parses the documents, stores the extracted content, and updates the visible insights after each upload.",
    ],
  },
  {
    heading: "System architecture",
    table: [
      ["Layer", "Role"],
      ["Frontend", "Handles upload, loading states, live views, chat, and export actions."],
      ["Backend", "Accepts PDFs, extracts text, builds the project model, and persists state."],
      ["AI layer", "Uses Gemini when a key is present, otherwise falls back to project-aware logic."],
      ["State store", "Keeps documents, messages, and insights available for every screen."],
    ],
    tableWidths: [120, 383],
  },
  {
    heading: "Why this is a working demo",
    bullets: [
      "Uploading a document changes the model, not just the file list.",
      "The UI now shows analyzing and insights-ready states during the flow.",
      "The same backend state powers the dashboard, Copilot, export PDF, and all analytics pages.",
    ],
  },
])

await createPdf("commissioning_assistant_workflow.pdf", "Commissioning Assistant Logic", [
  {
    heading: "Main use",
    paragraphs: [
      "This feature reads a commissioning test script or IST checklist and cross-references it against known open punch-list items or incomplete installs.",
      "The goal is to warn the team before testing starts, so a likely failure can be avoided instead of discovered late on site.",
    ],
  },
  {
    heading: "Example behavior",
    table: [
      ["Checklist item", "Result", "Reason"],
      ["Temporary power energization", "Likely fail", "UPS delay blocks the test."],
      ["Switchgear functional test", "Likely fail", "Incomplete install is still open."],
      ["BMS integration test", "At risk", "I/O verification is pending."],
      ["Fire alarm interface", "Pass", "No blocking issue found."],
    ],
    tableWidths: [170, 90, 243],
  },
  {
    heading: "Judge takeaway",
    bullets: [
      "This reduces rework by predicting failure before the test window starts.",
      "It is especially useful for commissioning managers and site engineers.",
    ],
  },
])

await createPdf("geospatial_tracking_workflow.pdf", "Geospatial Tracking Logic", [
  {
    heading: "Main use",
    paragraphs: [
      "This feature visualizes equipment shipment locations and status so the team can see which long-lead items are at risk and which downstream tasks are exposed.",
      "It is a polished visual layer, but it still feeds the same project risk engine as the other pages.",
    ],
  },
  {
    heading: "Example behavior",
    table: [
      ["Equipment", "Status", "Impact"],
      ["UPS", "In transit", "Delays energization and installation."],
      ["Transformer", "At dispatch hub", "May shift procurement-critical tasks."],
      ["Chiller", "On site", "No major downstream exposure."],
      ["Switchgear", "Customs clearance", "Can affect commissioning readiness."],
    ],
    tableWidths: [120, 120, 263],
  },
  {
    heading: "Judge takeaway",
    bullets: [
      "The map makes the supply chain easy to read at a glance.",
      "Its biggest value is the schedule warning attached to each shipment.",
    ],
  },
])

await createPdf("judge_demo_guide.pdf", "Judge Demo Guide", [
  {
    heading: "Best demo flow",
    table: [
      ["Action", "What the judge sees"],
      ["Upload a PDF", "A clear analyzing indicator appears."],
      ["Wait briefly", "The UI flips to insights ready."],
      ["Open Dashboard", "The project metrics update from the uploaded file."],
      ["Open Commissioning", "Likely failures are flagged automatically."],
      ["Open Copilot", "Ask a question and get a project-aware answer."],
      ["Export PDF", "A real PDF summary downloads from the backend."],
    ],
    tableWidths: [120, 263],
  },
  {
    heading: "Best framing",
    paragraphs: [
      "Tell the judges that the app is a connected decision-support prototype, not a static mockup.",
      "The value proposition is that one uploaded project PDF can update the entire operational picture.",
    ],
  },
])

console.log(`Generated PDFs in ${outDir}`)
